### PR TITLE
Librarian Merge Request Table

### LINKED ISSUE #4071: Librarian Merge Queue: Propose/Flag Dupes
**Adapt Integrated Librarian Environment (ILE) bar to propose Merge Requests**

mek  2:22 PM
We very briefly talked about how community might safely flag merge requests
@seabelis led us to a great idea, which is -- if not librarian, have same UI as if admin/librarian. Let people select items. Instead of showing "Merge Works" show "Merge Request"
Normally what happens is we send Lisa to /works/merge?works=...
Instead, clicking this link would take this URL and (instead of redirecting) just submit that URL to a review list, to lisa (edited) 

cdrini  2:24 PM
:plus1: That would be hugly wonderful/impactful, and not a ton of work! 

mek  2:24 PM
A bit messy and it's not something we should tackle this month given the inf. is in flux
But implementation wise, I think it's a good approach and simple to implement assuming everything is merged upstream

cdrini  2:24 PM
But I think we need to figure out the lists issue; otherwise Lisa will end up having queues of merges that she can't do

mek  2:25 PM
That's fine too, that's a different problem.
We just provide a sort or something. Or add another flag to e.g. "resolve" as unactionable. Or we compute that ahead of time and annotate the link accordingly. In any event, baby steps towards streamlining is better that what's happening today.

cdrini  2:26 PM
I don't think fixing the lists issue would be a ton of work, but it just needs to happen. After that, merges can happen at the speed of discovery :slightly_smiling_face:

### LINKED ISSUE #6715: Librarian Merge Queue: Create merge requests for Super Librarians
Ensure that a merge request is created for super librarians when they merge works with the ILE,

### PR DESCRIPTION
**Pull request by jimchamp**
Closes #4071
Closes #6715

This PR adds librarian merge queue functionality, namely:
1. Allows members of `librarian` and `admin` usergroups to submit a work merge request from a search result page.
2. Allows members of `librarian-work-merge` usergroup to update merge requests (MR) from the merge UI.

Members of group 2 (above) will be referred to as "Super Librarians" in this PR; members of group 1, "librarians".

This PR can be considered complete when the following tasks have been finished:
- [X] Work IDs in merge URL are naturally ordered.
- [X] Redundant work MRs are rejected.
- [X] Librarians can submit a MR with optional comment from work search pages.
- [X] Librarians are given a link to their entry on the `/merges` page on successful MR submission.
- [X] Super Librarians can reject a MR, with comments, from the merge UI.
- [X] MR is closed with optional comment when Super Librarian successfully merges works in MR.
- [ ] Handle cases where redirects appear in MRs (avoid race conditions) [#6675]
- [x] Handle merge UI making several unnecessary GET requests each time the text in the "Comments" input is updated.
- [x] Fix layout/styling of buttons + "Comments" input in merge UI.
- [x] Add "Open", "Closed", and "All" tabs to `/merges` page.
- [x] Improve styling of `/merges` table.
- [ ] Add pagination to `/merges` table.
- [ ] Ensure `community_edits_queue` submitters are covered by the patron anonymization process.

### Technical

We enforce a unique URL when creating a new merge request.  If the URL in the new request is already present in the database, we do not create a new row.  This should be fine for merge requests, but we will want to loosen this requirement for flagged records (if/when we include them in the queue).

### CODE REVIEW

**Comment by jimchamp** [comment]
Limitations to current feature:
- Comments cannot be edited nor deleted

Outstanding to-dos:
- [ ] Add pagination to request table views
- [ ] Add submission date to the views
- [ ] Trim comments before POSTing (in order to prevent empty string comments)
- [ ] Prompt submitters to only add clarifying comments to new requests (double-check copy of comment dialog)
- [ ] Create new merge request when super librarians select works and click "Do merge" in ILE bar

File: openlibrary/plugins/upstream/code.py [hunk 1]
```diff
@@ -99,14 +101,14 @@ class merge_work(delegate.page):
     path = "/works/merge"
 
     def GET(self):
+        i = web.input(records='', comment=None, mrid=None)
```
Comment (jimchamp) [inline openlibrary/plugins/upstream/code.py:104]:
```suggestion
        i = web.input(records='', mrid=None)
```
`comment` doesn't appear to be used here.

File: openlibrary/templates/merge_queue/comment.html [hunk 2]
```diff
@@ -0,0 +1,28 @@
+$def with(comment=None, comment_str=None)
```
Comment (jimchamp) [inline openlibrary/templates/merge_queue/comment.html:1]: Add top-level comments describing the difference between `comment` and `comment_str`.

File: openlibrary/templates/site/body.html [hunk 3]
```diff
@@ -6,8 +6,11 @@
 $ bodyattrs = ctx.get('bodyattrs', [])
 
 $if ctx.path.startswith('/works/OL') or ctx.path.startswith('/authors/OL') or ctx.path.startswith('/books/OL') or ctx.path.startswith('/search'):
-  $if ctx.user and ((ctx.user.is_librarian() or ctx.user.is_admin()) and ctx.user.is_usergroup_member('/usergroup/librarian-work-merge')):
+  $if ctx.user and ((ctx.user.is_librarian() or ctx.user.is_admin())):
     $bodyclass.append('show-librarian-tools')
+    $bodyattrs.append('data-username="%s"' % ctx.user.key.split('/')[-1])
+    $if ctx.user.is_usergroup_member('/usergroup/librarian-work-merge'):
+      $bodyattrs.append('data-can-merge="%s"' % 'true')
```
Comment (jimchamp) [inline openlibrary/templates/site/body.html:13]: This assumes that super librarians (members of `/usergroup/librarian-work-merge`) will also be a member of either `/usergroup/admin` or `/usergroup/librarians`.  If not, the super librarian will not have "Merge" links rendered in the table.

I'm also wondering if `admin`s should also be permitted to use the merge UI.  Currently, they can only submit merge requests.

**Review by mekarpeles — APPROVED** [review]

File: openlibrary/plugins/upstream/code.py (hunk 1 above)
Comment (jimchamp) [reply openlibrary/plugins/upstream/code.py:104]: `records` is used.
`comments` should be removed.

**Pull request merged**